### PR TITLE
Switch to markdown-it

### LINKED ISSUE #11: magnet links don't parse into anchor links in messages
likely `mailto:` will fail for similar reasons

to reproduce: 
1. wrap a valid magnet link as you would an http url
2. submit a message thru, say, patchwork

### LINKED ISSUE #12: Too eager to turn numbers into numbered lists
The following markdown:

``` markdown
> How many hops from my follows do I replicate by default?

3. I thought it was 2 (friends, foafs...
```

Was interpreted as the start of a numbered list. Wrong wrong wrong.

### LINKED ISSUE #13: closing parens are too greedily matched
see %Ds+A7fUbrqWNHssfyXUYq5GM1gkgmWh4GCwvkBzix4s=.sha256 for an example

### LINKED ISSUE #16: does this have injection attacks?
I look at the code and I'm not sure that this doesn't have attacks. given that this creates html on your computer from data distributed across a p2p network, if someone can create arbitary html (like script tags) by just posting code in a markdown link that would be a big security problem.

we need tests for this.

### LINKED ISSUE #19: # channel matching is including white space
In the message `%a5hjte9+fx1rR90yUGhJAsadv0c1TeVyZ+/kehgJCF0=.sha256` the author wrote "Type a # character in the search bar in the upper right to show the list of channels" and `# character` turned into a hyperlink. 

The white space after the `#` should cause this to be ignored, but it is not.  

cc @clehner

### PR DESCRIPTION
**Pull request by christianbundy**
Resolves #11 
Resolves #12 
Resolves #13
Resolves #16 ([xss test suite](https://github.com/markdown-it/markdown-it/blob/master/test/fixtures/markdown-it/xss.txt))
Resolves #19

---

**Is this useful/helpful/wanted?** I was wondering whether we could replace our fork of `marked` with an up-to-date version of [`markdown-it`](https://github.com/markdown-it/markdown-it), so this is a first stab at the implementation. If there's any interest, I'd like to continue to implement the rest of the functionality. Current caveats:

- ~~No respect for `opts`, are these worth implementing?~~
- ~~`magnet:` links are blacklisted specifically to pass tests~~
  - ~~Do we need to blacklist specific link types, or is there a whitelist?~~
    - ~~...or is the whitelist passed in `opts`? :grin:~~
- ~~No support for audio/video "images" in Markdown~~
- Had to modify tests because they expect non-standard whitespace behavior.
  - Old implementation was adding a space at the end of paragraphs.
  - Old implementation was treating `\n` as `<br>`, but this treats `\n\n` correctly as a paragraph break.
  - Old implementation replaces apostrophes with HTML entities. 
  - Old implementation left indentation spaces in inline code
    - Before:
      ```
      if (x) {   if (y) {     if (z) {       a()         .b()       }     }   }
      ```
    - After:
      ```
      if (x) { if (y) { if (z) { a() .b() } } }
      ```
- Had to modify tests because they expect non-standard fenced code blocks on inline.
  - The `lang-js` attribute was replaced with `language-js`, probably fixable but this is a spike.
    - **Update**: This is intentional and standard, more info here: https://spec.commonmark.org/0.21/#example-98
  - When you triple-backtick *and* set the language name, should the language name show inline text?
    - This branch assumes "yes" because there's no trivial way to do otherwise AFAIK.
    - **Update:** this is intentional, more info here: https://github.com/markdown-it/markdown-it/issues/146
- ~~No support for unlinked @mentions. If you want a mention link, you have to supply `[@mention](@link)`~~
  - ~~Was this being used? It seems like removing this feature reduces our complexity a *ton*.~~

This branch could probably use a refactor, too, but I made a bunch of progress and I wanted to get feedback before going further/deeper. Looking forward to hearing some thoughts! 

---

**Update summary:**

After pushing the new changes, it looks like the only changes from `master`'s behavior is:

- Standard Markdown whitespace behavior.
- Standard fenced code block `language-*` behavior.
- Inline fenced blocks are treated like inline fenced blocks, language isn't hidden.
- Unlinked @mentions are fully supported.

### CODE REVIEW

**Comment by dominictarr** [comment]
@christianbundy inline mentions are patchwork@2, hmm, also have a funny feeling they might be used in patchfoo?

are the "current caveats" describing your this PR so far, or are the limitations caused by markdown-it?

what are the benefits of switching to markdown-it? you've listed a bunch of problems so far.

**Comment by christianbundy** [comment]
@dominictarr Thanks for the quick feedback.

- Do those clients lack the `[@mention](@link)` option when composing a message? I don't think I've been able to grok how the current Markdown parser is finding users from arbitrary strings starting with `@`, especially names that contain spaces or other non-alphanum symbols.
- I think almost all of them are implementation caveats that should be trivial to fix, the only "restriction" is our non-standard whitespace: our fork is doing all sorts of funky things with whitespace, which markdown-it (and *probably* the current version of marked) aren't subject to.
- The problem being solved here is that our fork of marked is old and buggy and the merge conflict from upstream is no fun. I've tried to bring it up-to-date once before and it sucked, so I figured I'd just start from scratch and see what happens. Markdown-it also has drop-in plugins for hashtags and emojis, which made that part of the implementation easier. I don't know how many of these things are from ssb-markdown and how many are ssbc/marked, but here are some bugs I'd like to resolve:
  - `\n\n` should create a `<br>`, but our current implementation is treating `\n` as `<br>`
  - our current implementation can't handle links that end with parens
  - html entities are broken, `&amp;` is rendered as "[&amp](#) ;"
  - hashtags can end with colons, so `#feedback: ` is rendered as "[#feedback:](#)"

FWIW, I don't have any strong preferences between marked and markdown-it. Another alternative would be to switch to an up-to-date version of marked, but from my experience the four years of upgrades was basically a rewrite anyway.

**Comment by christianbundy** [comment]
Just added support for mentions without Markdown links and reorganized the tests so that `standard` didn't complain about backticks in our strings. As part of the reorganization I also made it so that the renderer is only instantiated once, which resulted in a huge performance boost. Should be good to go now.

File: lib/block.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+  Object.keys(media).forEach(key => {
+    const value = media[key]
+    if (value.length) {
+      properties += ` ${key}="${value}"`
+    }
+  })
+
+  if (alt.startsWith('audio:')) {
+    return `<video controls${properties}/>`
```
Comment (mmckegg) [inline lib/block.js]: this should probably be `<audio ...`

**Comment by christianbundy** [comment]
@dominictarr @mmckegg Is there anyone else I should get feedback from on this PR? I don't think Patchfoo uses ssb-markdown anymore, but I know Patchbay does.

cc: @mixmix

**Comment by mixmix** [comment]
This is great Christian,  im keen to smoke test it them merge it if its going well. 
We can fix things which are bung as we go. 

I'd anyone opposed to this plan? If they are we can make this into a module ssb-markdown-next !

File: test/input.json [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  },
+  {
+    "previous": "%S4N3qwNI/Uaf2U8DgsBhyhRJ5A6H4KFEEuwP7XF0SSA=.sha256",
+    "author": "@EMovhfIrFk4NihAKnRNhrfRaqIhBv1Wj8pTxJNgvCCY=.ed25519",
+    "sequence": 482,
+    "timestamp": 1451900123101,
+    "hash": "sha256",
+    "content": {
+      "type": "post",
+      "text": "@paul @mixmix in the latest version there isn't anyway to see who is on a thread until you see them reply. In previous versions the other recipients where in the reply form, but that is gone now. I can imagine some unfortunate things being said because someone doesn't realize who else is on that thread.",
```
Comment (mixmix) [inline test/input.json]: Haha,  this is a classic piece of history!

**Comment by mixmix** [comment]
Is the best way to test this npm linking it into patchcore?

I had noticed our anomoluous newline behavior. I kinda like it but given its non standard md we skills probably drop it

**Comment by christianbundy** [comment]
> Is the best way to test this npm linking it into patchcore?

Yeah, I used `npm link` for development but [my branch](https://github.com/ssbc/patchbay/tree/@christianbundy/master) uses:

```sh
npm i --save 'https://github.com/ssbc/ssb-markdown#markdown-it-implementation'
```

I think the anomalous newlines are slightly more intuitive, but they're awful when you use line breaks to wrap your text at ~80 characters. GitHub has the same anomaly, which makes commit messages and emails replies look terrible.

**Pull request merged**